amethyst_derive/src/prefab_data.rs

```diff
@@ -33,47 +36,216 @@ fn impl_prefab_data_component(ast: &DeriveInput) -> TokenStream {
+) -> (Vec<(Type, bool)>, TokenStream, TokenStream) {
+    let mut data_types = Vec::new();
+    let (add_to_entity, subs) = prepare_prefab_aggregate_fields(&mut data_types, &data.fields);
+    let extract_fields_add =
+        data.fields
+            .iter()
+            .enumerate()
+            .map(|(field_number, field)| match &field.ident {
+                Some(name) => quote! {
+                    let #name = &self.#name;
+                },
+                None => {
+                    let var_name =
+                        Ident::new(&format!("field_{}", field_number), Span::call_site());
+                    let number = Literal::usize_unsuffixed(field_number);
+                    quote! {
+                        let #var_name = &self.#number;
+                    }
+                }
+            });
+    let extract_fields_sub = data.fields.iter().enumerate().map(|(field_number, field)| {
+        if !have_component_attribute(&field.attrs[..]) {
+            match &field.ident {
+                Some(name) => Some(quote! {
+                    let #name = &mut self.#name;
+                }),
+                None => {
+                    let var_name =
+                        Ident::new(&format!("field_{}", field_number), Span::call_site());
+                    let number = Literal::usize_unsuffixed(field_number);
+                    Some(quote! {
+                        let #var_name = &mut self.#number;
+                    })
+                }
             }
+        } else {
+            None
         }
     });
-    let adds = (0..data.len()).map(|n| {
-        let tuple_index = Literal::usize_unsuffixed(n);
-        let (_, name, is_component) = &data[n];
+    (
+        data_types,
+        quote! {
+            #(#extract_fields_add)*
+            #(#add_to_entity)*
+        },
+        quote! {
+            #(#extract_fields_sub)*
+            #(#subs)*
+        },
+    )
+}
+
+fn prepare_prefab_aggregate_enum(
+    base: &Ident,
+    data: &DataEnum,
+) -> (Vec<(Type, bool)>, TokenStream, TokenStream) {
+    let mut data_types = Vec::new();
+    let mut subs = Vec::new();
+    let mut add_to_entity = Vec::new();
+
+    for variant in &data.variants {
+        let (variant_add_to_entity, variant_subs) =
+            prepare_prefab_aggregate_fields(&mut data_types, &variant.fields);
+        let field_names_add: Vec<_> = variant
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(field_number, field)| match &field.ident {
+                Some(name) => name.clone(),
+                None => Ident::new(&format!("field_{}", field_number), Span::call_site()),
+            })
+            .collect();
+        let field_names_sub: Vec<_> = variant
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(field_number, field)| match &field.ident {
+                Some(name) => {
+                    if !have_component_attribute(&field.attrs[..]) {
```

Comment:
this negation appears in a number of places, so this tiny helper function would help my brain a bit better 😛

```rust
#[inline]
fn is_aggregate_prefab(attrs: &[Attribute]) -> bool {
    !have_component_attribute(attrs)
}
```

Maybe rename `have_component_attribute` to `is_component_prefab`

Reply:
Legit.